sdks/go/pkg/beam/core/state/state.go

```diff
@@ -209,3 +224,162 @@ func MakeBagState[T any](k string) Bag[T] {
 		Key: k,
 	}
 }
+
+// CreateAccumulator represents a struct that implements the CreateAccumulator Combine operation
+type CreateAccumulator[T1 any] interface {
+	CreateAccumulator() T1
+}
+
+// AddInput represents a struct that implements the AddInput Combine operation
+type AddInput[T1, T2 any] interface {
+	AddInput(T1, T2) T1
+}
+
+// ExtractOutput represents a struct that implements the ExtractOutput Combine operation
+type ExtractOutput[T1, T3 any] interface {
+	ExtractOutput(T1) T3
+}
+
+// Combining is used to read and write global pipeline state representing a single combined value.
+// It uses 3 generic values, [T1, T2, T3], to represent the accumulator, input, and output types respectively.
+// Key represents the key used to lookup this state.
+type Combining[T1, T2, T3 any] struct {
+	Key     string
+	accumFn interface{}
+}
+
+// Add is used to write add an element to the combining pipeline state.
+func (s *Combining[T1, T2, T3]) Add(p Provider, val T2) error {
+	// We will always maintain a single accumulated value as a value state.
+	// Therefore, when we add we must first read the current accumulator so that we can add to it.
+	acc, ok, err := s.readAccumulator(p)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		// If no accumulator, that means that the CreateAccumulator function doesn't exist
+		// and our value is our initial accumulator.
+		return p.WriteValueState(Transaction{
+			Key:  s.Key,
+			Type: TransactionTypeSet,
+			Val:  val,
+		})
+	}
+
+	if ai := p.AddInputFn(s.Key); ai != nil {
+		var newVal interface{}
+		if f, ok := ai.(reflectx.Func2x1); ok {
+			newVal = f.Call2x1(acc, val)
+		} else {
+			newVal = f.Call([]interface{}{acc, val})[0]
+		}
+		return p.WriteValueState(Transaction{
+			Key:  s.Key,
+			Type: TransactionTypeSet,
+			Val:  newVal,
+		})
+	}
+	// If AddInput isn't defined, that means we must just have one accumulator type identical to the input type.
+	if ma := p.MergeAccumulatorsFn(s.Key); ma != nil {
+		var newVal interface{}
+		if f, ok := ma.(reflectx.Func2x1); ok {
+			newVal = f.Call2x1(acc, val)
+		} else {
+			newVal = f.Call([]interface{}{acc, val})[0]
+		}
+		return p.WriteValueState(Transaction{
+			Key:  s.Key,
+			Type: TransactionTypeSet,
+			Val:  newVal,
+		})
+	}
+
+	// Should be taken care of by previous validation
+	panic(fmt.Sprintf("MergeAccumulators must be defined on accumulator %v", s))
+}
+
+// Read is used to read this instance of global pipeline state representing a combiner.
+// When a value is not found, returns an empty list and false.
+func (s *Combining[T1, T2, T3]) Read(p Provider) (T3, bool, error) {
+	acc, ok, err := s.readAccumulator(p)
+	if !ok || err != nil {
+		var val T3
```

Comment:
since this function is returning an empty list when value is not found, do we expect users to specify type as `[]int`? or just `int` and then returning `[]T3` from here